src/sctools/encodings.py

```diff
@@ -0,0 +1,169 @@
+import random
+
+
+class Encoding:
+    """Abstract base class for DNA encodings"""
+
+    encoding_map = NotImplemented
+    decoding_map = NotImplemented
+    bits_per_base = NotImplemented
+
+    @classmethod
+    def encode(cls, bytes_encoded):
+        """encode a DNA bytes string
+
+        :param int bytes_encoded: encoded DNA string
+        :return bytes: decoded bytes DNA sequence
+        """
+        raise NotImplementedError
+
+    def decode(self, integer_encoded):
+        """decode an encoded string
+
+        :param int integer_encoded: encoded DNA string
+        :return bytes: decoded bytes DNA sequence
+        """
+        raise NotImplementedError
+
+    def gc_content(self, integer_encoded):
+        raise NotImplementedError
+
+
+class TwoBit(Encoding):
+    """Encode a DNA sequence using a 2-bit encoding.
+
+    Two-bit encoding uses 0 for an encoded nucleotide. As such, it cannot distinguish between
+    the end of sequence and trailing A nucleotides, and thus decoding these strings requires
+    knowledge of their length. As such, it is only appropriate for encoding fixed sequence
+    lengths
+
+    In addition, in order to encode in 2-bit, N-nucleotides must be randomized to one of A, C,
+    G, and T.
+
+    :param int sequence_length: number of nucleotides that are being encoded
+    """
+
+    def __init__(self, sequence_length):
+        self.sequence_length = sequence_length
+
+    class TwoBitEncodingMap:
+        """
+        custom class that mimics a read-only dict, but generates random nucleotides for ambiguous
+        nucleotides e.g. N
+        """
+        map_ = {ord('A'): 0, ord('C'): 1, ord('T'): 2, ord('G'): 3,
+                ord('a'): 0, ord('c'): 1, ord('t'): 2, ord('g'): 3}
+
+        def __getitem__(self, byte):
+            try:
+                return self.map_[byte]
+            except KeyError:
+                if byte != 78:
```

Comment:
Just to keep everything consistent, you might want `byte in (ord("N"), ord("n"))`